modules/ip6/datapath/ndp_ns_input.c

```diff
@@ -83,20 +86,62 @@ static uint16_t ndp_ns_input_process(
 			goto next;
 		}
 
-		opt = (struct icmp6_opt *)rte_pktmbuf_adj(mbuf, sizeof(*ns));
 		lladdr_found = icmp6_get_opt(
-			opt, rte_pktmbuf_pkt_len(mbuf), ICMP6_OPT_SRC_LLADDR, &lladdr
+			PAYLOAD(ns), rte_pktmbuf_pkt_len(mbuf), ICMP6_OPT_SRC_LLADDR, &lladdr
 		);
 
+		if (rte_ipv6_addr_is_unspec(&src)) {
+			// - If the IP source address is the unspecified address, the IP
+			//   destination address is a solicited-node multicast address.
+			ASSERT_NDP(rte_ipv6_addr_is_mcast(&dst));
+			// - If the IP source address is the unspecified address, there is
+			//   no source link-layer address option in the message.
+			ASSERT_NDP(!lladdr_found);
+			// If the source of the solicitation is the unspecified address, the
+			// node MUST set the Solicited flag to zero and multicast the
+			// advertisement to the all-nodes address.
+			src = (struct rte_ipv6_addr)RTE_IPV6_ADDR_ALLNODES_LINK_LOCAL;
+			solicited = false;
+		} else {
+			if (lladdr_found) {
+				// Copy the NS probe and send it to control plane for processing.
+				struct rte_mbuf *copy = rte_pktmbuf_copy(
+					mbuf, mbuf->pool, 0, UINT32_MAX
+				);
+				if (copy == NULL) {
+					next = ERROR;
+					goto next;
+				}
+				if (gr_mbuf_is_traced(mbuf)) {
+					gr_mbuf_trace_copy(copy, gr_mbuf_traces(mbuf));
+					uint8_t trace_len = RTE_MIN(
+						payload_len, GR_TRACE_ITEM_MAX_LEN
+					);
+					struct icmp6 *t = gr_mbuf_trace_add(mbuf, node, trace_len);
+					memcpy(t, icmp6, trace_len);
+				}
+				struct control_output_mbuf_data *d;
+				d = control_output_mbuf_data(copy);
+				d->iface = iface;
+				d->callback = ndp_probe_input_cb;
+				rte_node_enqueue_x1(graph, node, CONTROL, copy);
+			}
+			// Otherwise, the node MUST set the Solicited flag to one and unicast the
+			// advertisement to the Source Address of the solicitation.
+			solicited = true;
+		}
+
 		// Reuse the mbuf to forge a neighbour advertisement reply.
-		// XXX,TODO: avoid issues with encap, remember previous data offset?
+		offset = mbuf->data_off;
 		rte_pktmbuf_reset(mbuf);
+		mbuf->data_off = offset;
```

Comment:
Ugh!

Reply:
I removed this. Completely unrelated. I wonder if calling `rte_pktmbuf_reset()` is required after all.

Reply:
This seems more like a global scrub on the packet without knowing what was the reason.
I agree, rte_pktmbuf_reset() does not seem required.

Reply:
And that would partially address the concern I had about the data offset.